### PR TITLE
One getter for syncVersion, bust a potential cached syncversion where appropriate, save Changes if handleunsafemigrate makes changes

### PR DESCRIPTION
**Pull request by nick4598**

### CODE REVIEW

File: packages/transformer/src/test/TestUtils/IModelTestUtils.ts [hunk 1]
```diff
@@ -1117,6 +1130,68 @@ export class IModelTestUtils {
...
+      );
+    expect(scopeProvenanceCandidates).to.have.length(1);
+    const targetScopeProvenance =
+      scopeProvenanceCandidates[0].toJSON() as ExternalSourceAspectProps;
+    const targetScopeJsonProps = JSON.parse(
+      targetScopeProvenance.jsonProperties
+    ) as TargetScopeProvenanceJsonProps;
+
+    if (expectedProps.syncVersionId === undefined) {
+      const regex = new RegExp(`;${expectedProps.syncVersionIndex}$`);
```
Comment (JulijaRamoskiene) [inline packages/transformer/src/test/TestUtils/IModelTestUtils.ts:1151]: Q: Why this needs to be a regexp?

Comment (JulijaRamoskiene) [reply packages/transformer/src/test/TestUtils/IModelTestUtils.ts:1151]: Same question for reverseSyncVersion validation

File: packages/transformer/src/IModelTransformer.ts [hunk 2]
```diff
@@ -1038,11 +1012,17 @@ export class IModelTransformer extends IModelExportHandler {
...
-      if (!version) {
+      if (
+        !version &&
+        this._options.branchRelationshipDataBehavior === "unsafe-migrate"
+      ) {
         return { index: -1, id: "" }; // previous synchronization was done before fed guid update.
       }
-
-      const [id, index] = version.split(";");
+      if (version === undefined) {
```
[the remark this answers is not in the document]
Comment (nick4598) [reply packages/transformer/src/IModelTransformer.ts:1021]: If theres something wrong with the target scope ESA, I'd prefer for the user to sort of acknowledge that by having to opt into the unsafe-migrate behavior. It atleast gives them a chance to provide their own version / reversesyncversion or atleast explicitly acknowledge its unsafe and we'll give them whatever version we end up with after the transform is complete. I believe without throwing here, user could be none the wiser and get whatever version we end up with after the transform is complete.

File: packages/transformer/src/test/TestUtils/IModelTestUtils.ts (hunk 1 above)
Comment (nick4598) [reply packages/transformer/src/test/TestUtils/IModelTestUtils.ts:1151]: I added this as a helper function for some tests, in most cases when writing a test I don't know the changesetId so I will not pass it to this helper function. The regex allows us to assert that the index I passed is part of the sync /reversesyncversion.

Without regex I would always have to pass ID + index and then compare id;index to the actual id;index in the target ESA.

Maybe you're saying I could instead just do a endsWith comparison? (targetScopeProvenance.version?.endsWith(`;${syncVersionIndex}`)

File: packages/transformer/src/IModelTransformer.ts [hunk 3]
```diff
@@ -1146,39 +1128,59 @@ export class IModelTransformer extends IModelExportHandler {
       aspectProps.jsonProperties = foundEsaProps.jsonProperties
         ? JSON.parse(foundEsaProps.jsonProperties)
         : undefined;
-      this.handleUnsafeMigrate(aspectProps);
+      if (this.handleUnsafeMigrate(aspectProps)) {
+        // We only update the aspect if handleUnsafeMigrate made a change.
+        this.provenanceDb.elements.updateAspect({
```
[the remark this answers is not in the document]
Comment (nick4598) [reply packages/transformer/src/IModelTransformer.ts:1147]: added info log

File: packages/transformer/src/IModelTransformer.ts [hunk 4]
```diff
@@ -1146,39 +1128,59 @@ export class IModelTransformer extends IModelExportHandler {
...
+          ...aspectProps,
+          jsonProperties: JSON.stringify(aspectProps.jsonProperties) as any,
+        });
+        // Busting a potential cached version
+        this._cachedSynchronizationVersion = undefined;
```
[the remark this answers is not in the document]
Comment (nick4598) [reply packages/transformer/src/IModelTransformer.ts]: added private method

File: packages/transformer/src/IModelTransformer.ts [hunk 5]
```diff
@@ -3251,7 +3257,12 @@ export class IModelTransformer extends IModelExportHandler {
     }
     this._csFileProps = csFileProps;
 
-    this._sourceChangeDataState = "has-changes";
+    /** Theres a possibility that our csFileProps length is still 0 here, since we skip cs indices found in the pendingSync and pendingReverseSync indices arrays. */
+    if (this._csFileProps.length === 0) {
```
[the remark this answers is not in the document]
Comment (nick4598) [reply packages/transformer/src/IModelTransformer.ts]: switched to ternary

**Review by JulijaRamoskiene — APPROVED** [review]

File: packages/transformer/src/IModelTransformer.ts [hunk 6]
```diff
@@ -1146,39 +1136,65 @@ export class IModelTransformer extends IModelExportHandler {
...
         ? JSON.parse(foundEsaProps.jsonProperties)
         : undefined;
-      this.handleUnsafeMigrate(aspectProps);
+      // Clone oldProps incase they're changed for logging purposes
+      const oldProps = JSON.parse(JSON.stringify(aspectProps));
+      if (this.handleUnsafeMigrate(aspectProps)) {
+        Logger.logInfo(
+          loggerCategory,
+          "Unsafe migrate made a change to the target scope's external source aspect. Updating aspect in database.",
+          { oldProps, newProps: aspectProps }
```
[the remark this answers is not in the document]
Comment (nick4598) [reply packages/transformer/src/IModelTransformer.ts:1145]: Yes I did, but typescript will auto expand that into oldProps: oldProps

image
See above image for the eslint error I get if I were to do it that way. Both ways will log the same message

**Review by JulijaRamoskiene — APPROVED** [review]

**Pull request merged**